### PR TITLE
Create common ``Printer`` base class for ``pyreverse`` and improve typing.

### PR DESCRIPTION
**Pull request by DudeNr33**
## Steps

- [x] Write a good description on what the PR does.

## Description
This is the first MR which shall split MR #4615 into smaller and more easily reviewable parts.
It involves the necessary refactoring which is needed to prepare for the new features (PlantUML support and colorization).
It also adds type annotations.

Summary of changes:

* Rename ``vcgutils.py`` to ``printer.py`` and introduce a ``Printer`` base class which defines the common interface for the different backends.
* Adapt ``VCGPrinter`` to adhere to the new printer base class
* Add a ``DotPrinter`` class which essentially implements the features of ``pylint.graph.DotBackend``. The latter will be removed in a future MR, as it is currently also used by the ``ImportsChecker`` and I wanted to keep this MR smaller and focused (it is already not very small...).
* Add type annotations for the ``Printer`` class and its children. This also uncovered some problems where ``mypy`` correctly complained, so I had to add specialised ``PackageEntity`` and ``ClassEntity`` child classes in ``diagrams.py``.
* Added tests for VCG output, which was missing before
* Updated the test data as the new ``Printer`` backends now outputs some more information. The final image produced by Graphviz is still the same.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

File: pylint/pyreverse/utils.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/utils.py:280]: I think it's replaceable by ``pathlib.Path(filename_or_extension).suffix``

**Review by Pierre-Sassoulas — COMMENTED** [review]
Just a small comment, and I did not look in detail but the new typing is 🔥 , thanks! I'll review when this is no longer a draft :)

File: pylint/pyreverse/printer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/printer.py:57]:
```suggestion
class Printer(abc.ABC):
```
It's more explicit than raising ``NotImplementError``s :)

File: pylint/pyreverse/printer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/printer.py:111]: We can create a new file for VCGPrinter and another one for DotPrinter :) (It would also help with the diff and possible future conflict if vcgutils.py is recognized as the old file for it).

File: pylint/pyreverse/writer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/writer.py:133]: Maybe we could start with a first merge request that create the following architecture and move relevant part of the code without modifying anything:
```
pyreverse:
    writter:
        diagram_writter.py
        dot_writter.py
        vcg_writter.py
    printer:
        dot_printer.py
        vcg_printer.py
```
What do you think ?

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Hey, thank you for this work on this and sorry for the delay before reviewing. Could we reorganize and move the code around first then do the modification on it ? I'd be more confident if there was a proper diff for the change in the printer.py file, right now everything is considered new by git and it's hard to tell what changed exactly :)

**Comment by DudeNr33** [comment]
Thanks for the feedback. I will implement it when #4745 is merged and I did a rebase.

File: pylint/pyreverse/diagrams.py
Comment (DudeNr33) [inline pylint/pyreverse/diagrams.py:59]: Adding the type hints revealed that it was necessary to distinguish between a ``PackageEntity`` and a ``ClassEntity``, because the ``ClassEntity`` has additional attributes that were dynamically added in the previous code, which confused ``mypy``.

File: pylint/pyreverse/writer.py
Comment (DudeNr33) [inline pylint/pyreverse/writer.py:148]: ``get_values`` is now ``get_class_properties``.

``get_package_properties`` is currently trivial and the color is hardcoded, but this will change when the new feature for automatic coloring are merged.

File: pylint/pyreverse/writer.py
Comment (DudeNr33) [inline pylint/pyreverse/writer.py:187]: The ``__init__`` is now obsolete. It formerly defined the styles to apply to the different types of edges for displaying relationships.
These are now defined in the ``ARROWS`` dictionary in the module of the corresponding printer class.

**Comment by DudeNr33** [comment]
I implemented most of the feedback.
I reintroduced the module ``vcgutils.py``, which should make it now easier to diff.
As soon as the review is finished, I would suggest to make a last commit where this file is simply renamed to ``vcg_printer.py``.

Regarding the last suggestion (doing a first MR to implement the structure 
```
pyreverse:
    writter:
        diagram_writter.py
        dot_writter.py
        vcg_writter.py
    printer:
        dot_printer.py
        vcg_printer.py
```
):

If the current cleanup is not enough to properly review the changes, let me know and I can do that.

File: pylint/pyreverse/vcgutils.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/vcgutils.py:279]: 👍

File: pylint/pyreverse/writer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/writer.py:161]:
```suggestion
                    arguments: List[str] = [
```
Personal preference but when naming an iterable I like a plural word so I can do ``for argument in arguments`` later on.

File: pylint/pyreverse/dot_printer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/dot_printer.py:31]: Maybe we could create enums for fontcolor, arrowtail, style and arrowhead ?

File: pylint/pyreverse/dot_printer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/dot_printer.py:130]: We could create a function for this I think.

File: pylint/pyreverse/dot_printer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/dot_printer.py:34]: I think we can use ``Layout.LEFT_TO_RIGHT.name`` / ``Layout.LEFT_TO_RIGHT.value`` and remove this dict.

File: pylint/pyreverse/vcgutils.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/vcgutils.py:177]: Maybe a function that return the proper value in the Layout enum (Orientation or rankdir) would reduce the number of thing to keep in mind when using the code.

File: pylint/pyreverse/writer.py
Comment (Pierre-Sassoulas) [inline pylint/pyreverse/writer.py:60]:
```suggestion
        for module in sorted(diagram.modules(), key=lambda x: x.title):
```

File: pylint/pyreverse/writer.py
Comment (Pierre-Sassoulas) [reply pylint/pyreverse/writer.py:128]: Thank you for reorganizing the code the diff is manageable now :)

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for your work on this, very impressive ! Ad thank you for making the diff manageable. I left some minor comment. We can still do the architecture change later if you feel like it (probably after merging the other MR as the rebase will be hard to do if we do it before).

File: pylint/pyreverse/writer.py
Comment (DudeNr33) [reply pylint/pyreverse/writer.py:161]: I agree, changed it.

File: pylint/pyreverse/dot_printer.py
Comment (DudeNr33) [reply pylint/pyreverse/dot_printer.py:130]: Yes, good catch. ``main.py`` already does a similar check, they just differ in the consequence if ``dot`` is not found - one uses ``sys.exit(32)``, the other raises ``RuntimeError``. Settled for ``sys.exit(32)``.

File: pylint/pyreverse/dot_printer.py
Comment (DudeNr33) [reply pylint/pyreverse/dot_printer.py:34]: Yes, for the ``DotPrinter`` this is possible. Changed it.

File: pylint/pyreverse/vcgutils.py
Comment (DudeNr33) [reply pylint/pyreverse/vcgutils.py:177]: Hm, I don't know - we would just copy the content of the dictionaries into the ``Layout`` enum. Plus, the enum would then need to know the details of each backend (how the parameters have to look for ``dot``, ``vcg``, and ``puml`` - which is even more tricky as it only supports top to bottom and left to right, but not the other two). So every time a new backend is added, we would not only need to add the new printer class but also modify the ``Layout``.
This would be against the Open-Closed-Principle in my opinion.

File: pylint/pyreverse/writer.py
Comment (DudeNr33) [reply pylint/pyreverse/writer.py:60]: Yes, that's better. Changed it.

File: pylint/pyreverse/vcgutils.py
Comment (Pierre-Sassoulas) [reply pylint/pyreverse/vcgutils.py:177]: You're right I did not realize that :) !

**Comment by DudeNr33** [comment]
Thank you very much for your review! 
The only part I missed in the last push is the request for creating Enums for ``fontcolor``, ``arrowhead`` etc. 

I also need to rename the ``vcgutils.py`` into ``vcg_printer.py``, so please don't merge yet, will do that in the next few days!

**Comment by Pierre-Sassoulas** [comment]
No problem, just request a new review when the code is ready :)

**Comment by DudeNr33** [comment]
I thought about implementing enums for arrowstyle, arrowhead and so on a little longer.
In my opinion there is not much benefit from it in this case. We would have to create those enums for each printer separately, as the names and values for the arguments are different. And then they would only be used in the ``ARROWS`` dictionary, of which the items are not really individually processed:

```python
# in vcgutils.py
attributes = ARROWS[type_]
if label:
    attributes["label"] = label
self._write_attributes(
    EDGE_ATTRS,
    **attributes,
)

# in dot_printer.py
arrowstyle = ARROWS[type_]
attrs = [f'{prop}="{value}"' for prop, value in arrowstyle.items()]
```

I think we would just clutter the modules with enum definitions, without gaining much in terms of readability or reusability.

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for making a first step easier to review :) !

**Pull request merged**
